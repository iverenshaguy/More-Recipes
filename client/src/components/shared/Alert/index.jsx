import NormalAlert from './NormalAlert';

export default { NormalAlert };
