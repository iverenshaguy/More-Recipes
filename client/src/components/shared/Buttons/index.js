import AddRecipeBtn from './AddRecipeBtn';

export default { AddRecipeBtn };
