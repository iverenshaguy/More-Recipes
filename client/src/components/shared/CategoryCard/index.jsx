import React from 'react';
import { Link } from 'react-router-dom';
import { Card, CardImg, CardImgOverlay } from 'reactstrap';
import './CategoryCard.scss';

const CategoryCard = () => (
  <div className="col-xs-12 col-sm-6 col-md-4 col-lg-3 mt-4 category-card">
    <Link to="/">
      <Card className="m-0 p-0 card-image">
        <CardImg
          bottom
          width="100%"
          className="img-fluid"
          src="images/jollof-rice-img.jpg"
          alt="recipe"
        />
        <CardImgOverlay className="text-center card-image-overlay text-white px-2 recipe-rating">
          <h5 className="mb-1 pt-3">Category</h5>
        </CardImgOverlay>
      </Card>
    </Link>
  </div>
);

export default CategoryCard;
