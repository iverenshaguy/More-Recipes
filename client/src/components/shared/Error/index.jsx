import React from 'react';
import PropTypes from 'prop-types';
import './Error.scss';

/**
 * @exports
 * @function Error
 * @param {object} props
 * @returns {JSX} Error
 */
const Error = (props) => {
  const { type } = props;
  return (
    <div className="container align-items-center justify-content-center" id="error-wrapper">
      <div className="container">
        <h3 className="text-center">{type === 404 ? 'Page Not Found' : 'Something happened, please try again'}</h3>
      </div>
    </div>
  );
};

Error.propTypes = {
  type: PropTypes.number
};

Error.defaultProps = {
  type: 500
};

export default Error;
