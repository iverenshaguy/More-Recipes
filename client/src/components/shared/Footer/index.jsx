import React from 'react';
import './Footer.scss';

const Footer = () => (
  <footer className="footer">
    <div className="container">
      <p>Â© 2017 More Recipes</p>
    </div>
  </footer>
);

export default Footer;
