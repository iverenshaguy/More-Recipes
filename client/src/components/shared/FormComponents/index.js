import RenderInput from './RenderInput';
import RenderSelect from './RenderSelect';
import RenderFileInput from './RenderFileInput';
import RenderCheckBox from './RenderCheckBox';
import AddRecipeCompA from './AddRecipeCompA';
import AddRecipeCompB from './AddRecipeCompB';

export default {
  RenderInput,
  RenderSelect,
  RenderCheckBox,
  AddRecipeCompA,
  AddRecipeCompB,
  RenderFileInput
};
