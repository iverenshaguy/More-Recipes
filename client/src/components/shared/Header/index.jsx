import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Navbar from '../Navbar';
import { logout } from '../../../actions/auth';
import { toggleModal } from '../../../actions/ui';
import { userPropTypes } from '../../../helpers/proptypes';

/**
 * @exports Header
 * @class Header
 * @return {Component} Header
 */
class Header extends Component {
  /**
   * @memberof Header
   * @constructor
   * @returns {JSX} Logged out User
   */
  constructor() {
    super();
    this.logout = this.logout.bind(this);
    this.showAddRecipeModal = this.showAddRecipeModal.bind(this);
  }

  /**
   * @memberof Header
   * @returns {JSX} Logged out User
   */
  logout() {
    this.props.dispatch(logout());
  }

  /**
   * @memberof Header
   * @param {object} e
   * @returns {component} NavbarWrapper
   */
  showAddRecipeModal() {
    this.props.dispatch(toggleModal('addRecipe'));
  }

  /**
   * @memberof Header
   * @returns {JSX} Header
   */
  render() {
    return (
      <header>
        <Navbar

          isAuthenticated={this.props.isAuthenticated}
          logout={this.logout}
          showAddRecipeModal={this.showAddRecipeModal}
          user={this.props.user}
        />
      </header>
    );
  }
}

Header.propTypes = {
  isAuthenticated: PropTypes.bool.isRequired,
  dispatch: PropTypes.func.isRequired,
  ...userPropTypes
};

const mapStateToProps = state => ({
  isAuthenticated: state.auth.isAuthenticated,
  user: state.auth.user
});

export { Header as HeaderComponent };

export default connect(mapStateToProps)(Header);
