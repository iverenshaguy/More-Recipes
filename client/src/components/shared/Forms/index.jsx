import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Form as BootstrapForm, Button } from 'reactstrap';
import { asyncValidate, syncValidate } from '../../../helpers/validations';
import AuthForm from './AuthForm';
import { NormalAlert } from '../Alert';
import { arrayToObject } from '../../../utils';
import { formHelpers } from '../../../helpers';

/**
 * @exports
 * @class Form
 * @extends Component
 * @returns {component} Form
 */
class Form extends Component {
  static propTypes = {
    dispatch: PropTypes.func.isRequired,
    submitting: PropTypes.bool.isRequired,
    submitError: PropTypes.string,
    type: PropTypes.string.isRequired,
    meta: PropTypes.shape({
      title: PropTypes.string,
      btnText: PropTypes.string,
      extra: PropTypes.element
    }).isRequired
  };

  static defaultProps = {
    submitError: null
  }

  /**
   * @constructor
   * @memberof Form
   * @param {object} props - props
   * @returns {nothing} Return nothing
   */
  constructor(props) {
    super(props);
    const { type } = props;
    const { formFields } = formHelpers;
    const fields = formFields[type];

    this.state = {
      type,
      values: arrayToObject(fields, ''),
      touched: arrayToObject(fields, false),
      error: arrayToObject(fields, null),
      pristine: true,
      formValid: false,
      asyncValidating: false
    };

    this.handleBlur = this.handleBlur.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleFocus = this.handleFocus.bind(this);
    this.validateField = this.validateField.bind(this);
    this.validateForm = this.validateForm.bind(this);
  }

  /**
   * @memberof Form
   * @returns {nothing} Returns nothing
   */
  componentWillMount() {
    this.clearFormErrors();
  }

  /**
   * @memberof Form
   * @returns {nothing} Returns nothing
   */
  clearFormErrors() {
    const { type } = this.state;
    const { clearFormError } = formHelpers;

    this.props.dispatch(clearFormError[type]);
  }


  /**
   * @memberof Form
   * @param {object} event
   * @returns {nothing} Returns nothing
   */
  handleChange(event) {
    const { target } = event;
    const { value, name } = target;

    this.setState({
      values: { ...this.state.values, [name]: value },
      touched: { ...this.state.touched, [name]: true },
      pristine: false,
    }, () => { this.validateField(name, value); });
  }

  /**
   * @memberof Form
   * @returns {nothing} Returns nothing
   */
  handleFocus() {
    this.clearFormErrors();
  }

  /**
   * @memberof Form
   * @param {object} event
   * @returns {nothing} Returns nothing
   */
  handleBlur(event) {
    const { target } = event;
    const { value, name } = target;
    const { type } = this.state;

    this.setState({
      touched: { ...this.state.touched, [name]: true },
    }, () => { this.validateField(name); });

    setTimeout(() => {
      if ((name === 'email' || name === 'username') && this.state.error[name] === null) {
        // only set asyncvalidating when sync validation is completed
        this.setState({ asyncValidating: true }, () => {
          asyncValidate(type)(name, value)
            .then(() => { this.setState({ asyncValidating: false }); })
            .catch((error) => {
              this.setState({
                error: { ...this.state.error, [name]: error[name] },
                asyncValidating: false
              });
            });
        });
      }
    }, 500);
  }

  /**
   * @memberof Form
   * @param {string} field
   * @param {string} values
   * @returns {nothing} Returns nothing
   */
  validateField(field) {
    const { type } = this.state;
    const error = syncValidate(type)(field, this.state.values);

    if (error) {
      this.setState({
        error: {
          ...this.state.error,
          [field]: error
        }
      }, this.validateForm);
    } else {
      this.setState({
        error: {
          ...this.state.error,
          [field]: null
        }
      }, this.validateForm);
    }
  }

  /**
   * @memberof Form
   * @returns {nothing} Returns nothing
   */
  validateForm() {
    const formErrorArrayLength =
      Object.values(this.state.error).filter(value => value !== null).length;
    const formTouchedArrayLength =
      Object.values(this.state.touched).filter(value => value === true).length;
    const formFieldCount = Object.keys(this.state.values).length;

    if (formErrorArrayLength) {
      this.setState({
        formValid: false
      });
    } else if (!formErrorArrayLength && formTouchedArrayLength === formFieldCount) {
      this.setState({
        formValid: true
      });
    }
  }

  /**
   * @memberof Form
   * @param {object} event
   * @returns {nothing} Returns nothing
   */
  handleSubmit(event) {
    event.preventDefault();
    const { type } = this.state;
    const { formSubmitMapper } = formHelpers;

    this.props.dispatch(formSubmitMapper[type](this.state.values));
  }

  /**
   * @memberof Form
   * @returns {component} Form
   */
  render() {
    const {
      values, touched, error, pristine, formValid, asyncValidating
    } = this.state;

    const {
      submitting, submitError, type, meta
    } = this.props;

    const { title, btnText, extra } = meta;

    const formState = {
      values,
      touched,
      error,
      pristine,
      formValid,
      asyncValidating
    };

    const handlers = {
      handleChange: this.handleChange,
      handleBlur: this.handleBlur,
      handleFocus: this.handleFocus,
      handleSubmit: this.handleSubmit
    };

    return (
      <div>
        <h4 className="text-center">{title}</h4>
        <hr />
        {type !== 'login' &&
          <p className="text-muted mx-auto text-center">
            Fields marked
            <span className="text-danger">*</span> are important
          </p>}
        <BootstrapForm className="mt-4 mb-3 px-5" onSubmit={handlers.handleSubmit}>
          {submitError && (
            <NormalAlert color="danger">
              <p className="text-center mb-0">{submitError}</p>
            </NormalAlert>
          )}
          {(type === 'signup' || type === 'login') && <AuthForm type={type} state={formState} handlers={handlers} />}
          <Button className="btn-block mt-0" disabled={!formValid || pristine || submitting}>
            {btnText}
          </Button>
        </BootstrapForm>
        {extra && extra}
      </div>);
  }
}


export { Form as FormComponent };

export default connect()(Form);
