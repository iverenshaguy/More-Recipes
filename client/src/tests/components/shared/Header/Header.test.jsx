import React from 'react';
import { MemoryRouter } from 'react-router-dom';
// import Navbar from '../../Navbar';
import { HeaderComponent } from '../../../../components/shared/Header';

const dispatchMock = jest.fn();

const props = {
  isAuthenticated: false,
  currentLocation: 'auth',
  user: { firstname: 'Dave', lastname: 'Smith' },
  dispatch: dispatchMock
};

describe('Header', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<HeaderComponent {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });

  it('logs user out on logout click', () => {
    const component = mount( //eslint-disable-line
      <MemoryRouter>
        <HeaderComponent
          user={{ firstname: 'Dave', lastname: 'Smith' }}
          isAuthenticated
          currentLocation="auth"
          dispatch={dispatchMock}
        />
      </MemoryRouter>);
    const wrapper = component.find(HeaderComponent);

    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.find('a[href="/login"]').simulate('click', { preventDefault() { } });

    expect(dispatchMock).toHaveBeenCalled();
    component.unmount();
  });

  it('calls Modal on Add Recipe click', () => {
    const component = mount( //eslint-disable-line
      <MemoryRouter>
        <HeaderComponent
          user={{ firstname: 'Dave', lastname: 'Smith' }}
          isAuthenticated
          currentLocation="auth"
          dispatch={dispatchMock}
        />
      </MemoryRouter>);
    const wrapper = component.find(HeaderComponent);

    wrapper.find('a[href="#add-edit-modal"]').simulate('click', { preventDefault() { } });

    expect(wrapper.find('.modal')).toBeTruthy();
    component.unmount();
  });
});
