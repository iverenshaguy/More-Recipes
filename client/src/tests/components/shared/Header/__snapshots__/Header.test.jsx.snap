// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header logs user out on logout click 1`] = `
<Header
  currentLocation="auth"
  dispatch={[Function]}
  isAuthenticated={true}
  user={
    Object {
      "firstname": "Dave",
      "lastname": "Smith",
    }
  }
>
  <header>
    <NavbarWrapper
      isAuthenticated={true}
      logout={[Function]}
      showAddRecipeModal={[Function]}
      user={
        Object {
          "firstname": "Dave",
          "lastname": "Smith",
        }
      }
    >
      <Navbar
        className="navbar-expand"
        color="faded"
        expand={false}
        id="more-recipes-navbar"
        light={true}
        tag="nav"
      >
        <nav
          className="navbar-expand navbar navbar-light bg-faded"
          id="more-recipes-navbar"
        >
          <Link
            className="navbar-brand"
            replace={false}
            to="/"
          >
            <a
              className="navbar-brand"
              href="/"
              onClick={[Function]}
            >
              MORE RECIPES
            </a>
          </Link>
          <div
            className="collapse navbar-collapse justify-content-between"
            id="navbarNav"
          >
            <Nav
              className="ml-auto"
              navbar={true}
              tag="ul"
              vertical={false}
            >
              <ul
                className="ml-auto navbar-nav"
              >
                <UncontrolledDropdown
                  className="justify-content-start notifications-link"
                  nav={true}
                >
                  <Dropdown
                    className="justify-content-start notifications-link"
                    dropup={false}
                    isOpen={false}
                    nav={true}
                    toggle={[Function]}
                  >
                    <Manager
                      className="justify-content-start notifications-link dropdown nav-item"
                      onKeyDown={[Function]}
                      tag="li"
                    >
                      <li
                        className="justify-content-start notifications-link dropdown nav-item"
                        onKeyDown={[Function]}
                      >
                        <DropdownToggle
                          aria-haspopup={true}
                          color="secondary"
                          nav={true}
                        >
                          <Target
                            aria-expanded={false}
                            aria-haspopup={true}
                            className="nav-link"
                            component="a"
                            href="#"
                            onClick={[Function]}
                          >
                            <a
                              aria-expanded={false}
                              aria-haspopup={true}
                              className="nav-link"
                              href="#"
                              onClick={[Function]}
                            >
                              <FontAwesome
                                name="bell"
                                size="lg"
                              >
                                <span
                                  aria-hidden={true}
                                  className="fa fa-bell fa-lg"
                                />
                              </FontAwesome>
                            </a>
                          </Target>
                        </DropdownToggle>
                        <DropdownMenu
                          flip={true}
                          right={true}
                          tag="div"
                        >
                          <div
                            aria-hidden={true}
                            className="dropdown-menu dropdown-menu-right"
                            role="menu"
                            tabIndex="-1"
                          >
                            <div
                              className="container-fluid"
                            >
                              <p>
                                You have a new review
                              </p>
                              <div
                                className="dropdown-divider"
                              />
                              <p>
                                3 people viewed your recipe
                              </p>
                              <div
                                className="dropdown-divider"
                              />
                              <p>
                                Review your security details
                              </p>
                              <div
                                className="dropdown-divider"
                              />
                              <p>
                                Chef Steve just posted a new recipe
                              </p>
                              <div
                                className="dropdown-divider"
                              />
                              <p>
                                1 person added your recipe to their favorites
                              </p>
                            </div>
                          </div>
                        </DropdownMenu>
                      </li>
                    </Manager>
                  </Dropdown>
                </UncontrolledDropdown>
                <UncontrolledDropdown
                  nav={true}
                >
                  <Dropdown
                    dropup={false}
                    isOpen={false}
                    nav={true}
                    toggle={[Function]}
                  >
                    <Manager
                      className="dropdown nav-item"
                      onKeyDown={[Function]}
                      tag="li"
                    >
                      <li
                        className="dropdown nav-item"
                        onKeyDown={[Function]}
                      >
                        <div
                          className="nav-profile-picture-div rounded-cirle menu-dropdown d-inline"
                        >
                          <DropdownToggle
                            aria-haspopup={true}
                            color="secondary"
                            nav={true}
                          >
                            <Target
                              aria-expanded={false}
                              aria-haspopup={true}
                              className="nav-link"
                              component="a"
                              href="#"
                              onClick={[Function]}
                            >
                              <a
                                aria-expanded={false}
                                aria-haspopup={true}
                                className="nav-link"
                                href="#"
                                onClick={[Function]}
                              >
                                <img
                                  alt="User"
                                  className="rounded-cirle"
                                  src="images/user-image-placeholder.png"
                                />
                              </a>
                            </Target>
                          </DropdownToggle>
                          <DropdownMenu
                            flip={true}
                            right={true}
                            tag="div"
                          >
                            <div
                              aria-hidden={true}
                              className="dropdown-menu dropdown-menu-right"
                              role="menu"
                              tabIndex="-1"
                            >
                              <Link
                                className="dropdown-item"
                                replace={false}
                                to="/undefined"
                              >
                                <a
                                  className="dropdown-item"
                                  href="/undefined"
                                  onClick={[Function]}
                                >
                                  Profile
                                </a>
                              </Link>
                              <Link
                                className="dropdown-item"
                                replace={false}
                                to="/undefined/profile/edit"
                              >
                                <a
                                  className="dropdown-item"
                                  href="/undefined/profile/edit"
                                  onClick={[Function]}
                                >
                                  Edit Profile
                                </a>
                              </Link>
                              <div
                                className="dropdown-divider"
                              />
                              <a
                                className="dropdown-item"
                                href="/login"
                                onClick={[Function]}
                              >
                                Log out
                              </a>
                            </div>
                          </DropdownMenu>
                        </div>
                      </li>
                    </Manager>
                  </Dropdown>
                </UncontrolledDropdown>
                <UncontrolledDropdown
                  nav={true}
                >
                  <Dropdown
                    dropup={false}
                    isOpen={false}
                    nav={true}
                    toggle={[Function]}
                  >
                    <Manager
                      className="dropdown nav-item"
                      onKeyDown={[Function]}
                      tag="li"
                    >
                      <li
                        className="dropdown nav-item"
                        onKeyDown={[Function]}
                      >
                        <DropdownToggle
                          aria-haspopup={true}
                          className="menu-dropdown"
                          color="secondary"
                          nav={true}
                        >
                          <Target
                            aria-expanded={false}
                            aria-haspopup={true}
                            className="menu-dropdown nav-link"
                            component="a"
                            href="#"
                            onClick={[Function]}
                          >
                            <a
                              aria-expanded={false}
                              aria-haspopup={true}
                              className="menu-dropdown nav-link"
                              href="#"
                              onClick={[Function]}
                            >
                              <FontAwesome
                                name="reorder"
                                size="lg"
                              >
                                <span
                                  aria-hidden={true}
                                  className="fa fa-reorder fa-lg"
                                />
                              </FontAwesome>
                            </a>
                          </Target>
                        </DropdownToggle>
                        <DropdownMenu
                          flip={true}
                          right={true}
                          tag="div"
                        >
                          <div
                            aria-hidden={true}
                            className="dropdown-menu dropdown-menu-right"
                            role="menu"
                            tabIndex="-1"
                          >
                            <a
                              className="dropdown-item add-recipe"
                              href="#add-edit-modal"
                              onClick={[Function]}
                              title="New Recipe"
                            >
                              Add a Recipe
                            </a>
                            <Link
                              className="dropdown-item"
                              replace={false}
                              to="/"
                            >
                              <a
                                className="dropdown-item"
                                href="/"
                                onClick={[Function]}
                              >
                                Top Recipes
                              </a>
                            </Link>
                            <a
                              className="dropdown-item"
                              href="#favorite-recipes"
                              title="favorite-recipes"
                            >
                              Favorite Recipes
                            </a>
                          </div>
                        </DropdownMenu>
                      </li>
                    </Manager>
                  </Dropdown>
                </UncontrolledDropdown>
              </ul>
            </Nav>
          </div>
        </nav>
      </Navbar>
    </NavbarWrapper>
  </header>
</Header>
`;

exports[`Header renders correctly 1`] = `
<header>
  <NavbarWrapper
    isAuthenticated={false}
    logout={[Function]}
    showAddRecipeModal={[Function]}
    user={
      Object {
        "firstname": "Dave",
        "lastname": "Smith",
      }
    }
  />
</header>
`;
